lalrpop/src/rust/mod.rs

```diff
@@ -82,32 +92,48 @@ impl<'me, W: Write> RustWrite<W> {
                 if !first && session.emit_whitespace {
                     write!(self.write, " ")?;
                 }
-                write!(self.write, "{},", i)?;
+                write!(self.write, "{i},")?;
                 first = false;
             }
         }
         writeln!(self.write)
     }
 
-    pub fn writeln(&mut self, out: &str) -> io::Result<()> {
-        let buf = out.as_bytes();
+    // This function is intentionally not implemented with the `Write` trait
+    // because it is only called using the `rust!` macro, and thus with a `buf`
+    // that always ends with a `\n`.
+    #[doc(hidden)]
+    #[track_caller]
+    pub fn write_fmt(&mut self, args: fmt::Arguments<'_>) -> io::Result<()> {
+        // avoid intermediate allocation for simple format strings
+        let _storage;
+        let buf = match args.as_str() {
+            Some(s) => s.as_bytes(),
+            None => {
+                _storage = fmt::format(args);
+                _storage.as_bytes()
+            }
+        };
 
         // pass empty lines through with no indentation
-        if buf.is_empty() {
-            return self.write.write_all("\n".as_bytes());
+        match buf {
+            [] => return Ok(()),
+            [b'\n'] => return self.write.write_all(b"\n"),
+            [.., b'\n'] => {}
+            _ => unreachable!("rust! macro should always end with a newline"),
         }
 
-        let n = buf.len() - 1;
-
         // If the line begins with a `}`, `]`, or `)`, first decrement the indentation.
-        if buf[0] == b'}' || buf[0] == b']' || buf[0] == b')' {
+        if matches!(buf.first().unwrap(), b'}' | b']' | b')') {
```

Comment:
```suggestion
        if [b'}', b']', b')'].contains(buf.first().unwrap()) {
```

Reply:
It probably gets optimized to the same code, but why `array.contains` instead of a match?

Reply:
I guess so about the optimized result.

When I suggested it, I thought this is more like "one of collections" test rather than "pattern matching test". So suggested more typical expression for it.
But it may be regarded different for other people.

Reply:
As a datapoint, I think I prefer the pattern matching, or the initial test, because I feel like pattern matching or equality are more fundamental and common language constructs than slices and a call to the `contains` method. But it's mostly a question of taste, at this point.

Reply:
On the other hand, is `buf.first().unwrap()` much better? We unwrap anyway, and it just feels a bit more cluttering (also, we use `buf[n]` below, instead of `get(n).unwrap()`)

Reply:
Given that `buf.len() >= 2`, both index ops can never panic, though they should be written in the same style. I agree on the equality point, but personally I'd use pattern matching because it's much more concise and a very nice Rust language feature: `x, A | B | C` vs `x == A || x == B || x == C`.

Reply:
Yeah, I said equality because it's fine as well, but I also find the pattern matching to be the better as well, FWIW

Reply:
Thank you all. That sounds reasonable.